tuf/api/keys.py

```diff
@@ -0,0 +1,427 @@
+        if sys.version_info[0] >= 3 and isinstance(bytes_or_str, str):
+            return bytes_or_str.encode(encoding=encoding)
+        else:
+            return bytes_or_str
+
+    @abstractmethod
+    def sign(self, signed: BytesOrStr) -> Dict:
+        raise NotImplementedError
+
+    @abstractmethod
+    def verify(self, signed: BytesOrStr, signature: Dict) -> bool:
+        raise NotImplementedError
+
+Keys = List[Key]
+
+class KeyRing:
+
+    def __init__(self, threshold: Threshold, keys: Keys):
+        if len(keys) < threshold.least:
+            logging.warning(f'{len(keys)} < {threshold.least}')
+        if len(keys) > threshold.most:
+            logging.warning(f'{len(keys)} > {threshold.most}')
+        self.threshold = threshold
+        self.keys = keys
+
+# Specific types of keys, such as those in RAM, Hashicorp Vault,
+# AWS KMS (TODO), Azure Key Vault (TODO),
+# Google Cloud Key Management Service (TODO), or on HSMs (TODO).
+
+class RAMKey(Key):
+
+    # In practice, these are the only schemes used in py-TUF.
+    FileHandlers = {
+        'ecdsa-sha2-nistp256': import_ecdsa_privatekey_from_file,
+        'ed25519': import_ed25519_privatekey_from_file,
+        'rsassa-pss-sha256': import_rsa_privatekey_from_file,
+    }
+
+    def __init__(self, obj: Dict) -> None: # pylint: disable=super-init-not-called
+        self.__obj = obj
+
+    @classmethod
+    def read_from_file(
+        cls,
+        filename: str,
+        scheme: str,
+        passphrase: Optional[str] = None,
+        storage_backend: Optional[StorageBackendInterface] = None,
+    ) -> 'RAMKey':
+        handler = cls.FileHandlers.get(scheme)
+        if not handler:
+            return ValueError(scheme)
+        obj = handler(
+            filename,
+            password=passphrase,
+            storage_backend=storage_backend
+        )
+        return cls(obj)
+
+    @property
+    def keyid(self) -> str:
+        return self.__obj['keyid']
+
+    def sign(self, signed: BytesOrStr) -> Dict:
+        signed_bytes = self._encode(signed)
+        return create_signature(self.__obj, signed_bytes)
+
+    def _verify_rsa_signature(
+        self,
+        signed: BytesOrStr,
+        signature: Dict,
+        salt_length: int = SaltLength.PSSSaltLengthEqualsHash
+    ) -> bool:
+        sig = signature['sig']
+        sig = binascii.unhexlify(sig.encode('utf-8'))
+        scheme = self.__obj['scheme']
+        public = self.__obj['keyval']['public']
+        signed_bytes = self._encode(signed)
+        return verify_rsa_signature(
+            sig,
+            scheme,
+            public,
+            signed_bytes,
+            salt_length=salt_length
+        )
+
+    def verify(self, signed: BytesOrStr, signature: Dict) -> bool:
+        signed_bytes = self._encode(signed)
+        return verify_signature(self.__obj, signature, signed_bytes)
+
+class VaultKey(Key):
+
+    class AuthenticationError(Exception): pass
```

Comment:
Tiny nit, but perhaps don't nest this exception class or the following enums?

Having the former in a standalone `exceptions` module and the latter in a `models` (or similar) module makes it a bit easier to reuse them in other contexts.